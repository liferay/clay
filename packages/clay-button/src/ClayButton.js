import 'clay-icon';
import ClayComponent from 'clay-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayButton.soy.js';

/**
 * Metal Clay Button component.
 * @extends ClayComponent
 */
class ClayButton extends ClayComponent {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayButton.STATE = {
	/**
	 * Aria label attribute for the button element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	ariaLabel: Config.string(),

	/**
	 * Renders the button as a block element.
	 * @default false
	 * @instance
	 * @memberof ClayButton
	 * @type {?boolean}
	 */
	block: Config.bool().value(false),

	/**
	 * Data to add to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?object}
	 */
	data: Config.object(),

	/**
	 * The button disabled attribute.
	 * @default false
	 * @instance
	 * @memberof ClayButton
	 * @type {?boolean}
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * Icon to be rendered in the button.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	icon: Config.string(),

	/**
	 * Render ClayIcon in the ClayButton element.
	 * @default left
	 * @instance
	 * @memberof ClayButton
	 * @type {?string}
	 */
	iconAlignment: Config.oneOf(['left', 'right']).value('left'),

	/**
	 * Id to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	id: Config.string(),

	/**
	 * The label of the button content.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(html|string|undefined)}
	 */
	label: Config.any(),

	/**
	 * Flag to indicate if button should be monospaced.
	 * @default false
	 * @instance
	 * @memberof ClayButton
	 * @type {?bool}
	 */
	monospaced: Config.bool().value(false),

	/**
	 * The name attribute value of the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	name: Config.string(),

	/**
	 * The name that will make the CSS class name that will represent the
	 * button size.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	size: Config.oneOf(['sm']),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	spritemap: Config.string(),

	/**
	 * The css class that colors the button. Style `unstyled` is only for internal
	 * purposes.
	 * @default primary
	 * @instance
	 * @memberof ClayButton
	 * @type {?(bool|string|undefined)}
	 */
	style: Config.oneOfType([
		Config.bool(),
		Config.oneOf([
			'link',
			'primary',
			'secondary',
			'unstyled',
		]),
	]).value('primary'),

	/**
	 * The title attribute of the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	title: Config.string(),

	/**
	 * The type attribute value of the element.
	 * @default button
	 * @instance
	 * @memberof ClayButton
	 * @type {?string}
	 */
	type: Config.oneOf(['button', 'reset', 'submit']).value('button'),

	/**
	 * The value attribute value of the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayButton
	 * @type {?(string|undefined)}
	 */
	value: Config.string(),
};

defineWebComponent('clay-button', ClayButton);

Soy.register(ClayButton, templates);

export {ClayButton};
export default ClayButton;
