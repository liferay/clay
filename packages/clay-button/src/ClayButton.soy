{namespace ClayButton}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param? ariaLabel: string}
	{@param? block: bool}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? icon: string}
	{@param? iconAlignment: string}
	{@param? id: string}
	{@param? label: html|string}
	{@param? monospaced: bool}
	{@param? name: string}
	{@param? size: string}
	{@param? spritemap: string}
	{@param? style: bool|string}
	{@param? title: string}
	{@param? type: string}
	{@param? value: string}

	{let $attributes kind="attributes"}
		class="btn
			{if $block}
				{sp}btn-block
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}

			{if $monospaced}
				{sp}btn-monospaced
			{/if}

			{if $size}
				{sp}btn-{$size}
			{/if}

			{if isNonnull($style) and $style != true and $style != false}
				{sp}btn-{$style}
			{elseif not isNonnull($style) or $style == true}
				{sp}btn-primary
			{/if}
		"

		{if $ariaLabel}
			aria-label="{$ariaLabel}"
		{elseif $icon and not $label}
			aria-label="{$icon}"
		{/if}

		{if $disabled}
			disabled="disabled"
		{/if}

		{if $id}
			id="{$id}"
		{/if}

		{if $name}
			name="{$name}"
		{/if}

		{if $title}
			title="{$title}"
		{/if}

		{if $value}
			value="{$value}"
		{/if}

		type="{$type}"
	{/let}

	<button {$attributes}>
		{call .content}
			{param icon: $icon /}
			{param iconAlignment: $iconAlignment ?: 'left' /}
			{param label: $label /}
			{param spritemap: $spritemap /}
		{/call}
	</button>
{/template}

/**
 * Renders element content
 */
{template .content}
	{@param? icon: string}
	{@param? iconAlignment: string}
	{@param? label: html|string}
	{@param? spritemap: string}

	{let $iconContent kind="html"}
		{if $icon and $spritemap}
			{call .icon}
				{param icon: $icon /}
				{param iconAlignment: $iconAlignment /}
				{param label: $label /}
				{param spritemap: $spritemap /}
			{/call}
		{/if}
	{/let}

	{if $iconContent != '' and $iconAlignment == 'left'}
		{$iconContent}
	{/if}

	{$label ?: ''}

	{if $iconContent != '' and $iconAlignment == 'right'}
		{$iconContent}
	{/if}
{/template}

/**
 * Renders icon element
 */
{template .icon}
	{@param icon: string}
	{@param spritemap: string}
	{@param? iconAlignment: string}
	{@param? label: html|string}

	{let $iconContent kind="html"}
		{call ClayIcon.render}
			{param spritemap: $spritemap /}
			{param symbol: $icon /}
		{/call}
	{/let}

	{if $label}
		{let $iconWrapperClasses kind="text"}
			inline-item
			{if $iconAlignment == 'left'}
				{sp}inline-item-before
			{elseif $iconAlignment == 'right'}
				{sp}inline-item-after
			{/if}
		{/let}

		<span class="{$iconWrapperClasses}">
			{$iconContent}
		</span>
	{else}
		{$iconContent}
	{/if}

{/template}
