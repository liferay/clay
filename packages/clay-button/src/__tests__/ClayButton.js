import ClayButton from '../ClayButton';

let button;

describe('ClayButton', function() {
	afterEach(() => {
		if (button) {
			button.dispose();
		}
	});

	it('should render the default markup', function() {
		button = new ClayButton();

		expect(button).toMatchSnapshot();
	});

	it('should render a button with classes', function() {
		button = new ClayButton({
			elementClasses: 'my-custom-class',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with id', function() {
		button = new ClayButton({
			id: 'myId',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a submit button', function() {
		button = new ClayButton({
			type: 'submit',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a reset button', function() {
		button = new ClayButton({
			type: 'reset',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a disabled button', function() {
		button = new ClayButton({
			disabled: true,
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a secondary button', function() {
		button = new ClayButton({
			style: 'secondary',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with no style', function() {
		button = new ClayButton({
			style: false,
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with label', function() {
		button = new ClayButton({
			label: 'Label',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with name', function() {
		button = new ClayButton({
			name: 'myButton',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with label and ariaLabel', function() {
		button = new ClayButton({
			ariaLabel: 'My Description',
			label: 'Visit Liferay.com',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a small button', function() {
		button = new ClayButton({
			size: 'sm',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a block button', function() {
		button = new ClayButton({
			block: true,
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with icon', function() {
		button = new ClayButton({
			icon: 'plus',
			spritemap: 'icons.svg',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with icon and monospaced true', function() {
		button = new ClayButton({
			icon: 'plus',
			monospaced: true,
			spritemap: 'icons.svg',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with icon and label', function() {
		button = new ClayButton({
			icon: 'plus',
			label: 'Label',
			spritemap: 'icons.svg',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with label and icon', function() {
		button = new ClayButton({
			icon: 'plus',
			label: 'Label',
			spritemap: 'icons.svg',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with label and icon on right side', function() {
		button = new ClayButton({
			icon: 'plus',
			iconAlignment: 'right',
			label: 'Label',
			spritemap: 'icons.svg',
		});

		expect(button).toMatchSnapshot();
	});

	it('should render a button with customData attributes', function() {
		button = new ClayButton({
			data: {
				'my-attribute': 'myValue',
			},
		});

		expect(button.element.getAttribute('data-my-attribute')).toEqual(
			'myValue'
		);
	});
});
