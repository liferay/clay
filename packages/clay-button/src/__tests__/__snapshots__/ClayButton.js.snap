// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayButton should render a block button 1`] = `<button class="btn btn-block btn-primary" type="button"></button>`;

exports[`ClayButton should render a button with classes 1`] = `<button class="btn my-custom-class btn-primary" type="button"></button>`;

exports[`ClayButton should render a button with icon 1`] = `
<button class="btn btn-primary" aria-label="plus" type="button">
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
    <title>plus</title>
    <use xlink:href="icons.svg#plus"></use>
  </svg>
</button>
`;

exports[`ClayButton should render a button with icon and label 1`] = `
<button class="btn btn-primary" type="button">
  <span class="inline-item inline-item-before">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </span>Label</button>
`;

exports[`ClayButton should render a button with icon and monospaced true 1`] = `
<button class="btn btn-monospaced btn-primary" aria-label="plus" type="button">
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
    <title>plus</title>
    <use xlink:href="icons.svg#plus"></use>
  </svg>
</button>
`;

exports[`ClayButton should render a button with id 1`] = `<button class="btn btn-primary" id="myId" type="button"></button>`;

exports[`ClayButton should render a button with label 1`] = `<button class="btn btn-primary" type="button">Label</button>`;

exports[`ClayButton should render a button with label and ariaLabel 1`] = `<button class="btn btn-primary" aria-label="My Description" type="button">Visit Liferay.com</button>`;

exports[`ClayButton should render a button with label and icon 1`] = `
<button class="btn btn-primary" type="button">
  <span class="inline-item inline-item-before">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </span>Label</button>
`;

exports[`ClayButton should render a button with label and icon on right side 1`] = `
<button class="btn btn-primary" type="button">Label
  <span class="inline-item inline-item-after">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </span>
</button>
`;

exports[`ClayButton should render a button with name 1`] = `<button class="btn btn-primary" name="myButton" type="button"></button>`;

exports[`ClayButton should render a disabled button 1`] = `<button class="btn btn-primary" disabled="disabled" type="button"></button>`;

exports[`ClayButton should render a reset button 1`] = `<button class="btn btn-primary" type="reset"></button>`;

exports[`ClayButton should render a secondary button 1`] = `<button class="btn btn-secondary" type="button"></button>`;

exports[`ClayButton should render a small button 1`] = `<button class="btn btn-sm btn-primary" type="button"></button>`;

exports[`ClayButton should render a submit button 1`] = `<button class="btn btn-primary" type="submit"></button>`;

exports[`ClayButton should render the default markup 1`] = `<button class="btn btn-primary" type="button"></button>`;
